### ISSUE TITLE
New rule: `cast()` should always be a standalone assignment

### ISSUE THREAD

**Issue by KennethEnevoldsen** [astral-sh/ruff#25215]
### Summary

Add a lint rule that flags any `cast()` call that is not the sole right-hand side of a simple assignment statement.

**Motivation**
`cast()` is a type-system-only annotation with no runtime effect. When combined with other expressions, it obscures both the computation and the type assertion:

```py
# Bad
x = cast(MyType, some_function(arg1, arg2))
return cast("dict[str, Any]", json.loads(json.dumps(kwargs)))

# Good
x = some_function(arg1, arg2)
x = cast(MyType, x)

json_kwargs = json.loads(json.dumps(kwargs))
return cast("dict[str, Any]", json_kwargs)
```

---

It could also be argued that, e.g., `return cast(MyType, x)` should be flagged as the goal is to split functionality from typing information.

**Comment by MichaReiser** [comment]
`cast` isn't really a type assertion. In fact, it's the opposite. For example, this type checks without errors but is obviously wrong:

```py
a: int = cast(int, "test")
```

I don't think this changes your motivation, but I felt worth pointing out. To me, such a rule feels too strict, and it introduces the need for temporary variables (and shadowing). To me, the shadowing makes the code more obscure. That's why I'm not convinced that Ruff should add such a rule. But we can keep this issue open for a couple of weeks to to track user demand.

**Comment by KennethEnevoldsen** [comment]
Isn't `cast(t, x)` exactly a type assertion, it tells (assert to) the type checker that x is of a given type, t? However, (as you also point out), I think the point is mostly that it has no runtime effect - so more on the separation of concern.

I don't think introducing an extra variable is a big problem - e.g. here json_kwargs is a description of the output (a json conversion of the kwargs) which can often act as a replacement for a comment. 

While I see the issue with shadowing, I am not sure it outranks the separation of concerns.

I def. agree that the rule should be optional.

**Comment by jonathandung** [comment]
> Isn't `cast(t, x)` exactly a type assertion, it tells (assert to) the type checker that x is of a given type, t?

I think you meant asking the type checker to assume the type rather than to assert.

**Comment by KennethEnevoldsen** [comment]
I do think "assert" is the right word for what I meant (assert: to state a fact or belief confidently and forcefully), but I don't think the word choice makes the difference.

**Comment by tjkuson** [comment]
I think the terminology is significant seeing as both [`typing.cast`](https://docs.python.org/3/library/typing.html#typing.cast) and [`typing.assert_type`](https://docs.python.org/3/library/typing.html#typing.assert_type) exist and do distinct things.
